daemon/logger/awslogs/cloudwatchlogs.go

```diff
@@ -524,16 +530,17 @@ func (l *logStream) collectBatch(created chan bool) {
 // batch (defined in maximumBytesPerPut).  Log messages are split by the maximum
 // bytes per event (defined in maximumBytesPerEvent).  There is a fixed per-event
 // byte overhead (defined in perEventBytes) which is accounted for in split- and
-// batch-calculations.
-func (l *logStream) processEvent(batch *eventBatch, events []byte, timestamp int64) {
-	for len(events) > 0 {
+// batch-calculations.  Because the events are interpreted as UTF-8 encoded
+// Unicode, invalid UTF-8 byte sequences are replaced with the Unicode
+// replacement character (U+FFFD), which is a 3-byte sequence in UTF-8.  To
+// compensate for that and to avoid splitting valid UTF-8 characters into
+// invalid byte sequences, we calculate the length of each event assuming that
+// this replacement happens.
+func (l *logStream) processEvent(batch *eventBatch, bytes []byte, timestamp int64) {
+	for len(bytes) > 0 {
 		// Split line length so it does not exceed the maximum
-		lineBytes := len(events)
-		if lineBytes > maximumBytesPerEvent {
-			lineBytes = maximumBytesPerEvent
-		}
-		line := events[:lineBytes]
-
+		splitOffset, lineBytes := findValidSplit(string(bytes), maximumBytesPerEvent)
+		line := bytes[:splitOffset]
```

Comment:
If this comes back with both splitOffset and lineBytes as 0 do we end up adding a 0 length event incorrectly below during the batch.add call?

Reply:
I can add a check for that.  In practice, that shouldn't happen; there is no Unicode character that has a byte sequence greater than `maximumBytesPerEvent`, so there should always be a valid split if `bytes` is non-empty.

Reply:
Is bytes guaranteed to be non-empty?

Reply:
I'm going to leave this as-is, since we're just processing the `logger.Message` received by the log driver.  If `bytes` is empty, I think it's fine to generate an empty `cloudwatchlogs.InputLogEvent`.